TPC templates button

cristian-ungureanu opened the pull request:
### Summary
Added a new button to import a TPC template directly into a post, page, or custom layout

### Test instructions
- Go to a post, a page, or a custom layout, and you should be able to see that button
- The button is available only for PRO users, so make sure you activate the license
- Check if the import is working correctly. If you import more than one template, the adding should be done at the end.

### Screenshots or videos
https://vertis.d.pr/v/uowrr1

<!-- Issues that this pull request closes. -->
Closes Codeinwp/neve-pro-addon#2564.
<!-- Should look like this: `Closes #1, #2, #3.` . -->

HardeepAsrani commented on editor/src/components/import-modal.js:
```diff
@@ -0,0 +1,299 @@
+	);
+};
+
+export default compose(
+	withSelect( ( select ) => {
+		const { isPreview, getPreview, getCurrentTab } = select(
+			'tpc/block-editor'
+		);
+		return {
+			isPreview: isPreview(),
+			currentTab: getCurrentTab(),
+			previewData: getPreview().item,
+		};
+	} ),
+	withDispatch( ( dispatch ) => {
+		const { removeBlock, replaceBlocks, insertBlocks } = dispatch(
+			'core/block-editor'
+		);
+		const { togglePreview } = dispatch( 'tpc/block-editor' );
+
+		const closePreview = () => togglePreview( false );
+
+		return {
+			removeBlock,
+			replaceBlocks,
+			insertBlocks,
+			closePreview,
+		};
+	} )
+)( ImportModal );
```
NIT: We can use hooks instead of HOCs here.

HardeepAsrani commented on editor/src/index.js:
```diff
@@ -29,4 +31,8 @@ if ( parseInt( tiTpc.tier ) === 3 ) {
 		icon,
 	} );
+
+	registerPlugin( 'ti-tpc-templates-button', {
+		render: renderWrappedButton,
+	} );
```
Given Templates Patterns block import method is available to all users, we can also make it available to all users and not restrict it for Pro-only.

HardeepAsrani commented on editor/src/plugins/wrapped-tpc-templates-button.js:
```diff
@@ -0,0 +1,58 @@
+/**
+ * A wrapper for the TPC templates button that renders it inside post header toolbar.
+ */
+function WrappedTpcTemplatesButton() {
```
NIT: We can keep consistent and use arrow functions. And if there's only one export in the function, export default so you don't have to worry about the component naming during the import as I can see in case of this component.

HardeepAsrani reviewed (changes requested):
In terms of method, I agree with adding it to the bottom as you've done now.

For the design, I feel blue might have had been a better choice but we can ask @Codeinwp/design-team for their feelings on this.

Apart from this, I've left few smaller changes. Good work, overall!

mghenciu commented:
> For the design, I feel blue might have had been a better choice but we can ask https://github.com/orgs/Codeinwp/teams/design-team for their feelings on this.

In a separate context it may be better to have it blue, but overall I would say the the black could be more punchy(and we probably want this) in a context where there are multiple buttons active: 
![image](https://github.com/Codeinwp/templates-patterns-collection/assets/52494172/c65fc144-5f7d-48ba-8658-00d1c643bbf9)

cristian-ungureanu commented:
Thank you, everyone, for your two cents 🤗.

@HardeepAsrani, I've implemented the requested changes. I've changed the button position, and here's how it looks now, right next to the spectra icon.
<img width="1680" alt="Screenshot 2023-11-01 at 13 52 06" src="https://github.com/Codeinwp/templates-patterns-collection/assets/9929553/ce14c7c3-2b3f-489d-a866-4c89ffa85b52">

Personally, I would stick with the current black button because it's more evident, but I can change it in no time if we decide otherwise.

HardeepAsrani commented on editor/src/components/import-modal.js:
```diff
@@ -0,0 +1,270 @@
+import PreviewFrame from '../../../assets/src/Components/CloudLibrary/PreviewFrame';
+import { importTemplate } from '../data/templates-cloud';
+
+const { omit } = lodash;
```
We can dump lodash as 6.4 won't enqueue it by default, as similar issue being fixed in this PR: https://github.com/Codeinwp/templates-patterns-collection/pull/310

HardeepAsrani commented on editor/src/components/tpc-templates-button.js:
```diff
@@ -0,0 +1,25 @@
+			<Button
+				icon={ icon }
+				isPrimary
+				onClick={ () => modalState[ 1 ]( true ) }
```
While it should technically work, we can array destructuring for readability as we do everywhere else.

HardeepAsrani commented on editor/src/edit.js:
```diff
@@ -1,284 +1,14 @@
-		);
-	};
+const Edit = ( { autoLoad = true } ) => {
+	const modalState = useState( false );
```
NIT: Use array destructuring for useState.

HardeepAsrani reviewed (changes requested):
I left a couple of NITs but most importantly we need to avoid using lodash as it's been deprecated in the latest version of WordPress.

preda-bogdan reviewed (changes requested):
The code looks good, however I've left some changes for you to take care of and look into as to not break existing functionality and introduce any regressions.

Thank you!

cristian-ungureanu commented on editor/src/edit.js:
Thank you, @preda-bogdan, fixed all the issues.

irinelenache commented:
@cristian-ungureanu Tested and the feature works fine on my end ✅ 

I noticed that the button is available even when i use only Neve and TPC without license, but only the Page templates are available https://vertis.d.pr/i/FdQKve . I suppose this is the expected behaviour but i saw this in the test instructions _**The button is available only for PRO users, so make sure you activate the license**_

Let me know if that's expected or not 🙏

HardeepAsrani commented:
@irinelenache Yes, that's expected.
